doc/api/module.md

```diff
@@ -1048,6 +1048,73 @@ returned object contains the following keys:
 * columnNumber: {number} The 1-indexed columnNumber of the
   corresponding call site in the original source
 
+## TypeScript support
+
+<!-- YAML
+added: REPLACEME
+-->
+
+> Stability: 1.0 - Early development
+
+It is possible to execute TypeScript files by setting the experimental
+flag [`--experimental-strip-types`][].
+Node.js will replace inline TypeScript types with whitespace.
+No type checking is performed, and types are discarded.
+TypeScript features that cannot be replaced by whitespace, such as enums, will error.
+As in JavaScript files, file extensions are required in `import` statements
+and `import()` expressions.
+TypeScript features that depend on settings within `tsconfig.json`,
+such as paths or converting module formats, are unsupported.
+
+> The flag [`--experimental-require-module`][] is currently not supported.
+
+### Limitations
+
+#### Determining module system
+
+Node.js supports both [CommonJS][] and [ES Modules][] TypeScript syntax.
+When executing TypeScript files, Node.js will not transpile the module system.
+Since all that Node.js is doing is removing inline types, any `import`, `require`,
+`export` or `module.exports` statements, will be run as written.
+
+* `.ts` files will by default be as CommonJS modules,
+  unless specified otherwise, similarly to `.js` files.
+* `.mts` and `.cts` files will be treated as `.mjs` and `.cjs` files respectively.
+* `.tsx` files are not supported.
+
+Importing [modules without extension][] is not supported,
+file extensions are **always** required:
+
+```ts
+  import { foo } from './foo'; // it will not work
+  import { foo } from './foo.ts'; // it will work
+```
+
+The `tsconfig.json` option
+`allowImportingTsExtensions` may help provide compatibility with other tools.
```

Comment:
I think a pretty major caveat worth noting is that [`allowImportingTsExtensions` prevents `tsc` from compiling](https://www.typescriptlang.org/tsconfig/#allowImportingTsExtensions:~:text=this%20flag%20is%20only%20allowed%20when%20--noemit%20or%20--emitdeclarationonly%20is%20enabled) so this setup is mutually exclusive with compiling with `tsc` unless TypeScript decides to support import rewriting in the future.

Also, seems like there are more configs that should be recommended for TS to behave consistently with the runtime:

- [`"module": "nodenext" or "preserve"`](https://www.typescriptlang.org/tsconfig/#module) (and [`moduleResolution`](https://www.typescriptlang.org/tsconfig/#moduleResolution)) for Node.js features and resolution

- [`"allowJs": true`](https://www.typescriptlang.org/tsconfig/#allowJs) for supporting JS files to be imported from TS files, and supporting TS resolution from JS files

- [`"resolveJsonModule": true`](https://www.typescriptlang.org/tsconfig/#resolveJsonModule) for resolving `.json` files

- [`"verbatimModuleSyntax": true`](https://www.typescriptlang.org/tsconfig/#verbatimModuleSyntax) so type imports/exports can be stripped. This implicitly enables [`isolatedModules`](https://www.typescriptlang.org/tsconfig/#isolatedModules) which introduces constraints in how you write TS

- [`"moduleDetection": "force"`](https://www.typescriptlang.org/tsconfig/#moduleDetection) so each file is defaults to being treated as a module

Reply:
Maybe it makes sense to have a Node.js-recommended `tsconfig.json`

Reply:
This is something we want to discuss with the typescript team, but yeah 100% will be done

Reply:
If this PR lands, I think ideally there would be two new options added to `tsc`:

- `"allowTransforms": false` which would cause `tsc` to error on syntax that results in JavaScript code generation. Having `tsc` optionally error on syntaxes like enums would provide hints in editors so that authors are aware of the invalid syntax as soon as they write it, not later when they try to run it.

- `"rewriteTsExtensions": "js"` or similar, as a complement to `allowImportingTsExtensions`, where `tsc` replaces `.ts` at the end of imports with `.js` for files that `tsc` is compiling. I know this might be a hard sell to the TypeScript team as they’ve taken a firm position over the years that `tsc` won’t do any import specifier rewriting, but the result of that stance is that other tools are doing it instead. Many people like using `tsc` as their only build tool, so it would be a shame if users can’t continue to use it alongside `--strip-types`. I don’t think it should be Node’s responsibility to handle something that a build tool could and arguably should, since the build tool is what’s generating the new files and rewriting their contents.

Reply:
Knowing whether a module specifier *actually* refers to a `.ts` file, vs is a non-file specifier, is extremely fraught, and not an answerable question under the constraints that in-situ singe-file transpilers have.

Looking at a bare specifier `foo/bar.ts`, there are two possibilities:
 * `foo` is a package with a `package.json` that says `{ exports: { "bar.ts": something } }`, in which case rewriting this specifier will break it
 * `foo` is a path mapping to a local folder with an actual `bar.ts` file. In this case, you "must" rewrite it

Notably, during single-file-transpile, you don't even know what the config is or what the relevant package.jsons *are*, so there's no way to resolve this question.

It's too fraught to break into jail to do this. Bundlers are the right level of abstraction for transforming import specifiers into host-appropriate semantics.

Reply:
> not an answerable question under the constraints that in-situ singe-file transpilers have.

When you refer to “in-situ single file transpilers” do you have Node in mind or `tsc` in mind, or something else? In this PR, Node doesn’t rewrite any specifiers.

I don’t think rewriting `.ts` to `.js` is an unsolvable problem for `tsc`. It could do so incautiously, where if `"rewriteTsExtensions": "js"` is set then every `.ts`-ending specifier that’s relative or starting with a `tsconfig.json` `paths` mapping or starting with a `package.json` `#imports` mapping gets written out as `.js`, without checking for the existence of the source `.ts` file; or it could read the disk to be more careful. It could also do the rewriting only if the referenced file is one that `tsc` will also be transpiling. Sure there are complications and edge cases, but I think they’re all resolvable.